README.md

```diff
@@ -43,114 +43,280 @@
-      "FeaturevisorSDK"
-    ]
-  )
-)
+var options: InstanceOptions = .default
+options.datafileUrl = "https://cdn.yoursite.com/production/datafile-tag-all.json" 
+let featurevisor = try createInstance(options: options)
 ```
 
-Building your application will fetch and install everything locally:
+If you need to take further control on how the datafile is fetched, you can pass a custom `handleDatafileFetch` function
 
+```swift
+public typealias DatafileFetchHandler = (_ datafileUrl: String) -> Result<DatafileContent, Error>
 ```
-$ swift build
+
+```swift
+import FeaturevisorSDK
+
+var options: InstanceOptions = .default
+options.handleDatafileFetch = { datafileUrl in
+    ...
+}
+let featurevisor = try createInstance(options: options)
 ```
 
-## Usage
+### Context
+Contexts are a set of attribute values that we pass to SDK for evaluating features.
 
-@TODO: ...
+They are objects where keys are the attribute keys, and values are the attribute values.
+
+```swift
+public enum AttributeValue {
+    case string(String)
+    case integer(Int)
+    case double(Double)
+    case boolean(Bool)
+    case date(Date)
+```
+
+```swift
+let context = [
+  "myAttributeKey": .string("myStringAttributeValue"),
+  "anotherAttributeKey": .double(0.999),
+]
+```
+
+### Checking if enabled
+Once the SDK is initialized, you can check if a feature is enabled or not:
+
+```swift
+let featureKey = "my_feature";
+let context = [
+    "userId": .string("123"),
+    "country": .string("nl")
+]
+
+let isEnabled = sdk.isEnabled(featureKey: featureKey, context: context)
+```
+
+### Getting variations
+If your feature has any variations defined, you can get evaluate them as follows
+
+```swift
+let featureKey = "my_feature";
+let context = [
+    "userId": .string("123")
+]
+
+let variation = sdk.getVariation(featureKey: featureKey, context: context)
+```
+
+### Getting variables
+```swift
+let featureKey = "my_feature";
+let variableKey = "color"
+let context = [
+    "userId": .string("123")
+]
+sdk.getVariable(featureKey: featureKey, variableKey: variableKey, context: context)
+```
+
+### Type specific methods
+
+#### Boolean
+```swift
+sdk.getVariableBoolean(featureKey: FeatureKey, variableKey: VariableKey, context: Context)
+```
+
+#### String
+```swift
+sdk.getVariableString(featureKey: FeatureKey, variableKey: VariableKey, context: Context)
+```
+
+#### Integer
+```swift
+sdk.getVariableInteger(featureKey: FeatureKey, variableKey: VariableKey, context: Context)
+```
+
+#### Double
+```swift
+sdk.getVariableDouble(featureKey: FeatureKey, variableKey: VariableKey, context: Context)
+```
+
+#### Array of strings
+```swift
+sdk.getVariableArray(featureKey: FeatureKey, variableKey: VariableKey, context: Context)
+```
+
+#### Generic decodable object
+```swift
+sdk.getVariableObject(featureKey: FeatureKey, variableKey: VariableKey, context: Context)
+```
+
+#### JSON object
+```swift
+sdk.getVariableJSON(featureKey: FeatureKey, variableKey: VariableKey, context: Context)
```

Comment:
does Swift support generics? and if yes, should this method accept a type as a generic for the expected returned output?

like:

```swift
sdk.getVariableJSON<MyTypeHere>(...)
```

Reply:
Yep, we have generics :). Something like below is enough for the compiler to know the type:
`let variable: MyAwesomeType = sdk.getVariableJSON(featureKey: "feature_key", variableKey: "variable_key", context: context)`
Will update the readme to be more precise.